README
ocaml-speex


This package contains an O'Caml interface for 
the Speex codec library, otherwise known as speex.

Please read the COPYING file before using this software.

Prerequisites:
==============

- ocaml >= 3.0.6 (haven't tried earlier versions)

- speex >= 1.2 (haven't tried earlier versions)

- findlib >= 0.8.1 (haven't tried earlier versions)

Compilation:
============

	$ make all

This should build both the native and the byte-code version of the
extension library.

Installation:
=============

	$ make install

This should install the library file (using ocamlfind) in the
appropriate place.

Author:
=======

This author of this software may be contacted by electronic mail
at the following address: savonet-users@lists.sourceforge.net.


CHANGES
0.4.2 (unreleased)
=====
* Deprecate `Decoder.eos`

0.4.1 (04-01-2022)
=====
* Get rid of naked pointers.

0.4.0 (08-03-2021)
=====
* Switch to dune

0.3.0 (16-10-2020)
=====
* Switch to Unix.read style API.

0.2.2 (16-10-2020)
=====
* Fix Speex header notification spam.

0.2.1 (03-08-2015)
=====
* Changed types to match new ocaml-ogg's API.
  Should be fully backward compatible.
* Wrapped more speex functions around enter/leave
  blocking sections to avoid deadlock when speex
  writes to stderr

0.2.0 (04-07-2011)
=====
* Added decoding module for the new Ogg_demuxer module
* Implement new Out_of_sync exception
* Efficient seeking support

0.1.2 (14-10-2009)
=====
* Added support for --enable-debugging configure option
* Makefile now honnors LIBDIRS
  variable for linking against libraries
  located in other places than then standard
  ones.
* Added prefix to main compilation variables
  if passed to configure.
* Added NO_CUSTOM to build
  in standard mode.

0.1.1 (30-03-2009)
=====
* Fixed Makefile for BSD: call $(MAKE) for generating documentation.

0.1.0 (17-02-2009)
=====
* Initial release


examples/speex2wav.ml
(*
 * Copyright 2008 Savonet team
 *
 * This file is part of OCaml-speex.
 *
 * OCaml-speex is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * OCaml-speex is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with OCaml-speex; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *)

(**
  * A speex to wav converter using OCaml-speex.
  *
  * @author Samuel Mimram
  * @author Romain Beauxis
  *)

let src = ref ""
let dst = ref ""

let output_int chan n =
  output_char chan (char_of_int ((n lsr 0) land 0xff));
  output_char chan (char_of_int ((n lsr 8) land 0xff));
  output_char chan (char_of_int ((n lsr 16) land 0xff));
  output_char chan (char_of_int ((n lsr 24) land 0xff))

let output_short chan n =
  output_char chan (char_of_int ((n lsr 0) land 0xff));
  output_char chan (char_of_int ((n lsr 8) land 0xff))

let usage = "usage: speex2wav source destination"
let () = Speex_decoder.register ()

let _ =
  Arg.parse []
    (let pnum = ref (-1) in
     fun s ->
       incr pnum;
       match !pnum with
         | 0 -> src := s
         | 1 -> dst := s
         | _ ->
             Printf.eprintf "Error: too many arguments\n";
             exit 1)
    usage;
  if !src = "" || !dst = "" then (
    Printf.printf "%s\n" usage;
    exit 1);
  let dec, fd = Ogg_decoder.init_from_file !src in
  let { Ogg_decoder.audio_track; _ } = Ogg_decoder.get_standard_tracks dec in
  let audio_track =
    match audio_track with
      | None ->
          Printf.eprintf "Error: no audio track\n";
          exit 1
      | Some audio_track -> audio_track
  in
  let { Ogg_decoder.channels; sample_rate }, (encoder, comments) =
    Ogg_decoder.audio_info dec audio_track
  in
  Printf.printf "Encoder: %s\n" encoder;
  let print_comment (k, v) = Printf.printf "%s: %s\n" k v in
  Printf.printf "Comments:\n";
  List.iter print_comment comments;
  Printf.printf "\n";
  Printf.printf "Input file characteristics: %d channels, %d Hz\n" channels
    sample_rate;
  (* Using speex to decode the ogg. *)
  Printf.printf "\nDecoding...\n";
  flush_all ();
  let tmpdst, oc =
    Filename.open_temp_file ~mode:[Open_binary] "speex2wav" ".raw"
  in
  (try
     while true do
       Ogg_decoder.decode_audio dec audio_track (fun data ->
           let s1 = data.(0) in
           Array.iteri
             (fun n _ ->
               Array.iter
                 (fun s ->
                   let sample = int_of_float (s.(n) *. 32767.) in
                   output_short oc sample)
                 data)
             s1)
     done
   with Ogg.End_of_stream -> close_out oc);
  Printf.printf "Decoding finished, writing WAV..\n";
  Unix.close fd;
  (* Do the wav stuff. *)
  let datalen = (Unix.stat tmpdst).st_size in
  let ic = open_in_bin tmpdst in
  let oc = open_out_bin !dst in
  output_string oc "RIFF";
  output_int oc (4 + 24 + 8 + datalen);
  output_string oc "WAVE";
  output_string oc "fmt ";
  output_int oc 16;
  output_short oc 1;
  (* WAVE_FORMAT_PCM *)
  output_short oc 2;
  (* channels *)
  output_int oc sample_rate;
  (* freq *)
  output_int oc (sample_rate * 2 * 2);
  (* bytes / s *)
  output_short oc (2 * 2);
  (* block alignment *)
  output_short oc 16;
  (* bits per sample *)
  output_string oc "data";
  output_int oc datalen;
  (let buflen = 256 * 1024 in
   let buf = Bytes.create buflen in
   let r = ref 1 in
   let pos = ref 0 in
   while !r <> 0 do
     r := input ic buf 0 buflen;
     output oc buf 0 !r;
     pos := !pos + !r
   done);
  close_in ic;
  close_out oc;
  Unix.unlink tmpdst;
  Printf.printf "Done !\n";
  Gc.full_major ()


examples/wav2speex.ml
(*
 * Copyright 2008 Savonet team
 *
 * This file is part of ocaml-speex.
 *
 * ocaml-speex is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * ocaml-speex is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with ocaml-speex; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *)

(**
  * An wav to ogg converter using ocaml-speex.
  *
  * @author Samuel Mimram, Romain Beauxis, and many others...
  *)

open Speex
open Unix


let src = ref ""
let dst = ref ""

let input_string chan len =
  let ans = Bytes.create len in
    really_input chan ans 0 len ;
    Bytes.to_string ans

let input_int chan =
  let buf = input_string chan 4 in
    (int_of_char buf.[0])
    + (int_of_char buf.[1]) lsl 8
    + (int_of_char buf.[2]) lsl 16
    + (int_of_char buf.[3]) lsl 24

let input_short chan =
  let buf = input_string chan 2 in
    (int_of_char buf.[0]) + (int_of_char buf.[1]) lsl 8

let usage = "usage: wav2speex [options] source destination"
let float = ref false
let mode =  ref Wideband
let fpp = ref 5
let vbr = ref false
let quality = ref 7

let _ =
  let f = Printf.sprintf in
  let string_of_mode m = 
    match m with
      | Narrowband     -> "narrowband"
      | Wideband       -> "wideband"
      | Ultra_wideband -> "ultra-wideband"
  in
  Arg.parse
    [
      "--float", Arg.Unit (fun b -> float := true ),
      f "Use floats for decoding. Default: %b" !float;
      "--mode", Arg.String 
        (fun b -> 
          match b with
            | "narrowband"     -> mode := Narrowband
            | "wideband"       -> mode := Wideband
            | "ultra-wideband" -> mode := Ultra_wideband
            | _ -> failwith "unkown mode"),
      f "Encoding mode, one of \"narrowband\", \"wideband\" or \"ultra-wideband\". \
         Default: %s" (string_of_mode !mode);
      "--frame_per_packet", Arg.Int (fun b -> fpp := b ),
      f "Frames per Ogg packet. Default: %i" !fpp;
      "--vbr", Arg.Unit (fun b -> vbr := true ),
      f "Encode in vbr mode. Default: %b" !vbr;
      "--quality", Arg.Int (fun b -> quality := b ),
      f "Encoding bitrate, in Kbps. Default: %i" !quality;
    ]
    (
      let pnum = ref (-1) in
        (fun s -> incr pnum; match !pnum with
           | 0 -> src := s
           | 1 -> dst := s
           | _ -> Printf.eprintf "Error: too many arguments\n"; exit 1
        )
    ) usage;
  if !src = "" || !dst = "" then
    (
      Printf.printf "%s\n" usage;
      exit 1
    );
  let ic = open_in_bin !src in
  let oc = open_out_bin !dst in
    (* TODO: improve! *)
    if input_string ic 4 <> "RIFF" then invalid_arg "No RIFF tag";
    ignore (input_string ic 4);
    if input_string ic 4 <> "WAVE" then invalid_arg "No WAVE tag";
    if input_string ic 4 <> "fmt " then invalid_arg "No fmt tag";
    let _ = input_int ic in
    let _ = input_short ic in (* TODO: should be 1 *)
    let channels = input_short ic in
    let infreq = input_int ic in
    let _ = input_int ic in (* bytes / s *)
    let _ = input_short ic in (* block align *)
    let ibits = input_short ic in
    if input_string ic 4 <> "data" then invalid_arg "No data tag";
    let _ = input_int ic in (* datalen *)
    let fos buf =
      let len = String.length buf / (2 * channels) in
      let ans = Array.init channels (fun _ -> Array.make len 0) in
        for i = 0 to len - 1 do
          for c = 0 to channels - 1 do
            let n =
              int_of_char buf.[2 * channels * i + 2 * c]
              + int_of_char buf.[2 * channels * i + 2 * c + 1] lsl 8
            in
            let n =
              if n land 1 lsl 15 = 0 then
                n
              else
                (n land 0b111111111111111) - 32768
            in
             ans.(c).(i) <- max (-32768) (min 32767 n);
          done;
        done;
        ans
    in
    let enc = Encoder.init !mode !fpp in
    if not !vbr then
      Encoder.set enc SPEEX_SET_QUALITY !quality
    else
      Encoder.set enc SPEEX_SET_VBR_QUALITY !quality;
    Encoder.set enc SPEEX_SET_SAMPLING_RATE infreq;
    let ivbr = 
      if !vbr then 1 else 0 
    in
    Encoder.set enc SPEEX_SET_VBR ivbr;
    let fsize = Encoder.get enc SPEEX_GET_FRAME_SIZE in
    let os = Ogg.Stream.create () in
    let header = Header.init ~rate:infreq ~nb_channels:channels 
                             ~mode:(!mode) ~vbr:(!vbr) 
                             ~frames_per_packet:(!fpp) 
                             () in
    Header.encode_header header [] os;
    let s_o_f (h,b) = h ^ b in
    let flush s =
      let rec f v =
        try
          let v = v ^ (s_o_f (Ogg.Stream.flush_page s)) in
          f v
        with
          | Ogg.Not_enough_data -> v
      in
      f ""
    in
    output_string oc (flush os);
    let start = Unix.time () in
    let smode = 
      match !mode with
        | Narrowband     -> "narrowband"
        | Wideband       -> "wideband"
        | Ultra_wideband -> "ultra-wideband"
    in
      Printf.printf
        "Input detected: PCM WAVE %d channels, %d Hz, %d bits\n%!"
        channels infreq ibits;
      Printf.printf
        "Encoding to: SPEEX %d channels, %d Hz, %s, VBR: %s\nPlease wait...\n%!"
        channels infreq smode (string_of_bool !vbr);
        begin try while true do
          let buflen = 2*fsize*channels in
          let buf = Bytes.create buflen in
          let feed () =
            really_input ic buf 0 buflen;
            let buf = Bytes.to_string buf in
            let fbuf = fos buf in
            assert(Array.length (fbuf.(0)) = fsize);
            fbuf
          in
          let (h,v) =
            if channels = 1 then
              let feed () = 
                let frame = feed () in
                frame.(0)
              in
              if !float then
                let feed () = 
                  Array.map float_of_int (feed ())
                in 
                Encoder.encode_page enc os feed
              else
                Encoder.encode_page_int enc os feed   
            else
              if !float then
                let feed () = 
                  Array.map (fun x -> Array.map float_of_int x) (feed ())
                in
                Encoder.encode_page_stereo enc os feed
              else
                Encoder.encode_page_int_stereo enc os feed
          in
          output_string oc (h ^ v);
        done;
        with End_of_file -> () end;
        Encoder.eos enc os;
        output_string oc (flush os);
        close_in ic;
        close_out oc;
        Printf.printf "Finished in %.0f seconds.\n" ((Unix.time ())-.start);
        Gc.full_major ()


src/ogg_demuxer_speex_decoder.ml
(*
 * Copyright 2003-2011 Savonet team
 *
 * This file is part of Ocaml-speex.
 *
 * Ocaml-speex is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * Ocaml-speex is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Ocaml-speex; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *)

let check p =
  try
    let _ = Speex.Header.header_of_packet p in
    true
  with _ -> false

let decoder os =
  let dec_p = ref None in
  let decoder = ref None in
  let com_p = ref None in
  let os = ref os in
  let init () =
    match !decoder with
      | None ->
          let dec_p =
            match !dec_p with
              | None ->
                  let packet = Ogg.Stream.get_packet !os in
                  dec_p := Some packet;
                  packet
              | Some p -> p
          in
          let com_p =
            match !com_p with
              | None ->
                  let packet = Ogg.Stream.get_packet !os in
                  com_p := Some packet;
                  packet
              | Some p -> p
          in
          let header = Speex.Header.header_of_packet dec_p in
          let meta = Speex.Header.comments_of_packet com_p in
          let mode = header.Speex.Header.mode in
          let dec = Speex.Decoder.init mode in
          let sample_freq = header.Speex.Header.rate in
          let chans = header.Speex.Header.nb_channels in
          Speex.Decoder.set dec Speex.SPEEX_SET_SAMPLING_RATE sample_freq;
          decoder := Some (dec, sample_freq, chans, meta);
          (dec, sample_freq, chans, meta)
      | Some d -> d
  in
  let info () =
    let _, rate, chans, meta = init () in
    ({ Ogg_demuxer.channels = chans; sample_rate = rate }, meta)
  in
  let decode feed =
    let dec, _, chans, _ = init () in
    let len = ref 0 in
    let feed buf =
      let buf = Array.map (Array.map (fun x -> float x /. 32768.)) buf in
      len := !len + Array.length buf.(0);
      feed buf
    in
    try
      let decode dec os feed =
        if chans = 2 then Speex.Decoder.decode_int_feed_stereo dec os feed
        else (
          let feed x = feed [| x |] in
          Speex.Decoder.decode_int_feed dec os feed )
      in
      decode dec !os feed
    with Ogg.Not_enough_data -> if !len = 0 then raise Ogg.Not_enough_data
  in
  let restart new_os =
    os := new_os;
    let d, _, _, _ = init () in
    Speex.Decoder.set d Speex.SPEEX_RESET_STATE 0
  in
  Ogg_demuxer.Audio
    {
      Ogg_demuxer.name = "speex";
      info;
      restart;
      decode;
      samples_of_granulepos = (fun x -> x);
    }

let register () = Hashtbl.add Ogg_demuxer.ogg_decoders "speex" (check, decoder)


src/speex.ml
(*
 * Copyright 2003-2011 Savonet team
 *
 * This file is part of Ocaml-speex.
 *
 * Ocaml-speex is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * Ocaml-speex is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Ocaml-speex; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *)

(**
  * Functions for decoding speex files using libspeex.
  *
  * @author Romain Beauxis
  *)

exception Invalid_frame_size

type mode = Narrowband | Wideband | Ultra_wideband

(* Internal use only *)
type internal_mode

external internal_mode_of_int : int -> internal_mode = "caml_speex_get_mode"

(* Values from speex.h *)
let mode_of_int x =
  match x with
    | 0 -> Narrowband
    | 1 -> Wideband
    | 2 -> Ultra_wideband
    | _ -> failwith "unknown mode"

let int_of_mode x =
  match x with Narrowband -> 0 | Wideband -> 1 | Ultra_wideband -> 2

let internal_mode_of_mode x = internal_mode_of_int (int_of_mode x)

(* Generated by control_define *)

(** TODO: better implementation.. *)
type control =
  | SPEEX_RESET_STATE
  | SPEEX_SET_ENH
  | SPEEX_GET_ENH
  | SPEEX_GET_FRAME_SIZE
  | SPEEX_SET_QUALITY
  | SPEEX_SET_MODE
  | SPEEX_GET_MODE
  | SPEEX_SET_LOW_MODE
  | SPEEX_GET_LOW_MODE
  | SPEEX_SET_HIGH_MODE
  | SPEEX_GET_HIGH_MODE
  | SPEEX_SET_VBR
  | SPEEX_GET_VBR
  | SPEEX_SET_VBR_QUALITY
  | SPEEX_GET_VBR_QUALITY
  | SPEEX_SET_COMPLEXITY
  | SPEEX_GET_COMPLEXITY
  | SPEEX_SET_BITRATE
  | SPEEX_GET_BITRATE
  | SPEEX_SET_SAMPLING_RATE
  | SPEEX_GET_SAMPLING_RATE
  | SPEEX_SET_VAD
  | SPEEX_GET_VAD
  | SPEEX_SET_ABR
  | SPEEX_GET_ABR
  | SPEEX_SET_DTX
  | SPEEX_GET_DTX
  | SPEEX_SET_SUBMODE_ENCODING
  | SPEEX_GET_SUBMODE_ENCODING
  | SPEEX_SET_PLC_TUNING
  | SPEEX_GET_PLC_TUNING
  | SPEEX_SET_VBR_MAX_BITRATE
  | SPEEX_GET_VBR_MAX_BITRATE
  | SPEEX_SET_HIGHPASS
  | SPEEX_GET_HIGHPASS
  | SPEEX_GET_ACTIVITY

let int_of_control x =
  match x with
    | SPEEX_SET_ENH -> 0
    | SPEEX_GET_ENH -> 1
    | SPEEX_GET_FRAME_SIZE -> 3
    | SPEEX_SET_QUALITY -> 4
    | SPEEX_SET_MODE -> 6
    | SPEEX_GET_MODE -> 7
    | SPEEX_SET_LOW_MODE -> 8
    | SPEEX_GET_LOW_MODE -> 9
    | SPEEX_SET_HIGH_MODE -> 10
    | SPEEX_GET_HIGH_MODE -> 11
    | SPEEX_SET_VBR -> 12
    | SPEEX_GET_VBR -> 13
    | SPEEX_SET_VBR_QUALITY -> 14
    | SPEEX_GET_VBR_QUALITY -> 15
    | SPEEX_SET_COMPLEXITY -> 16
    | SPEEX_GET_COMPLEXITY -> 17
    | SPEEX_SET_BITRATE -> 18
    | SPEEX_GET_BITRATE -> 19
    | SPEEX_SET_SAMPLING_RATE -> 24
    | SPEEX_GET_SAMPLING_RATE -> 25
    | SPEEX_RESET_STATE -> 26
    | SPEEX_SET_VAD -> 30
    | SPEEX_GET_VAD -> 31
    | SPEEX_SET_ABR -> 32
    | SPEEX_GET_ABR -> 33
    | SPEEX_SET_DTX -> 34
    | SPEEX_GET_DTX -> 35
    | SPEEX_SET_SUBMODE_ENCODING -> 36
    | SPEEX_GET_SUBMODE_ENCODING -> 37
    | SPEEX_SET_PLC_TUNING -> 40
    | SPEEX_GET_PLC_TUNING -> 41
    | SPEEX_SET_VBR_MAX_BITRATE -> 42
    | SPEEX_GET_VBR_MAX_BITRATE -> 43
    | SPEEX_SET_HIGHPASS -> 44
    | SPEEX_GET_HIGHPASS -> 45
    | SPEEX_GET_ACTIVITY -> 47

let _ =
  Callback.register "caml_speex_mode_of_int" mode_of_int;
  Callback.register "caml_speex_int_of_mode" int_of_mode;
  Callback.register_exception "ocaml_speex_invfrlen_exn" Invalid_frame_size;
  Callback.register_exception "ocaml_speex_eos_exn" Ogg.End_of_stream

module Header = struct
  type t = {
    id : string;
    version : string;
    version_id : int;
    header_size : int;
    rate : int;
    mode : mode;
    mode_bitstream_version : int;
    nb_channels : int;
    bitrate : int;
    frame_size : int;
    vbr : bool;
    frames_per_packet : int;
    extra_headers : int;
  }

  (* Defined in speex_header.h *)
  let header_string_length = 8
  let header_version_length = 20

  external init : int -> int -> internal_mode -> int -> bool -> t
    = "caml_speex_init_header"

  let init ?(frames_per_packet = 1) ?(mode = Wideband) ?(vbr = true)
      ~nb_channels ~rate () =
    init rate nb_channels (internal_mode_of_mode mode) frames_per_packet vbr

  external encode_header_packetout :
    t -> string array -> Ogg.Stream.packet * Ogg.Stream.packet
    = "caml_speex_encode_header"

  let encode_header_packetout e l =
    let l = List.map (fun (x, y) -> Printf.sprintf "%s=%s" x y) l in
    encode_header_packetout e (Array.of_list l)

  let encode_header e l s =
    let p1, p2 = encode_header_packetout e l in
    Ogg.Stream.put_packet s p1;
    Ogg.Stream.put_packet s p2

  external header_of_packet : Ogg.Stream.packet -> t
    = "caml_speex_header_of_packet"

  external comments_of_packet : Ogg.Stream.packet -> string array
    = "caml_speex_comments_of_packet"

  let comments_of_packet p =
    let x = comments_of_packet p in
    let vendor = x.(0) in
    let x = Array.sub x 1 (Array.length x - 1) in
    let c_k = ref 0 in
    let c_v = ref 0 in
    let split s =
      try
        let i = String.index s '=' in
        try (String.sub s 0 i, String.sub s (i + 1) (String.length s - i - 1))
        with Invalid_argument _ ->
          c_v := !c_v + 1;
          (String.sub s 0 i, Printf.sprintf "unknown_value_%i" !c_k)
      with Not_found ->
        c_k := !c_k + 1;
        (Printf.sprintf "unkown_key_%i" !c_k, s)
    in
    (vendor, Array.to_list (Array.map split x))
end

module Encoder = struct
  type t

  external init : internal_mode -> int -> t = "ocaml_speex_enc_init"

  let init m x = init (internal_mode_of_mode m) x

  external get : t -> int -> int = "ocaml_speex_encoder_ctl_get"

  let get x q = get x (int_of_control q)

  external set : t -> int -> int -> unit = "ocaml_speex_encoder_ctl_set"

  let set x q v = set x (int_of_control q) v

  external encode_page_main :
    t -> int -> Ogg.Stream.stream -> (unit -> float array) -> Ogg.Page.t
    = "ocaml_speex_encode_page"

  let encode_page e s f = encode_page_main e 1 s f

  let merge frame =
    let ret =
      Array.init (2 * Array.length frame.(0)) (fun _ -> frame.(0).(0))
    in
    Array.iteri
      (fun i _ ->
        ret.(2 * i) <- frame.(0).(i);
        ret.((2 * i) + 1) <- frame.(1).(i))
      frame.(0);
    ret

  let encode_page_stereo e s f =
    let f () = merge (f ()) in
    encode_page_main e 2 s f

  external encode_page_int_main :
    t -> int -> Ogg.Stream.stream -> (unit -> int array) -> Ogg.Page.t
    = "ocaml_speex_encode_page_int"

  let encode_page_int e s f = encode_page_int_main e 1 s f

  let encode_page_int_stereo e s f =
    let f () = merge (f ()) in
    encode_page_int_main e 2 s f
end

module Decoder = struct
  type t

  external init : internal_mode -> t = "ocaml_speex_dec_init"

  let init m = init (internal_mode_of_mode m)

  external get : t -> int -> int = "ocaml_speex_decoder_ctl_get"

  let get x q = get x (int_of_control q)

  external set : t -> int -> int -> unit = "ocaml_speex_decoder_ctl_set"

  let set x q v = set x (int_of_control q) v

  let split frame =
    let ret = Array.make 2 (Array.make (Array.length frame / 2) frame.(0)) in
    for i = 0 to (Array.length frame / 2) - 1 do
      ret.(0).(i) <- frame.(2 * i);
      ret.(1).(i) <- frame.((2 * i) + 1)
    done;
    ret

  external decode_feed :
    t -> int -> Ogg.Stream.stream -> (float array -> unit) -> unit
    = "ocaml_speex_decoder_decode"

  let decode_gen e s chan func split =
    let l = ref [] in
    let feed x = l := split x :: !l in
    begin
      try func e chan s feed
      with Ogg.Not_enough_data ->
        if List.length !l = 0 then raise Ogg.Not_enough_data
    end;
    List.rev !l

  let decode e s = decode_gen e s 1 decode_feed (fun x -> x)
  let decode_stereo e s = decode_gen e s 2 decode_feed split

  let decode_feed_stereo e s feed =
    let feed x = feed (split x) in
    decode_feed e 2 s feed

  let decode_feed e s feed = decode_feed e 1 s feed

  external decode_int_feed :
    t -> int -> Ogg.Stream.stream -> (int array -> unit) -> unit
    = "ocaml_speex_decoder_decode_int"

  let decode_int e s = decode_gen e s 2 decode_int_feed (fun x -> x)
  let decode_int_stereo e s = decode_gen e s 2 decode_int_feed split

  let decode_int_feed_stereo e s feed =
    let feed x = feed (split x) in
    decode_int_feed e 2 s feed

  let decode_int_feed e s feed = decode_int_feed e 1 s feed
end

module Wrapper = struct
  module Decoder = struct
    exception Not_speex
    exception Internal

    type t =
      (Decoder.t
      * Ogg.Stream.stream
      * Ogg.Sync.t
      * nativeint
      * int
      * (string * (string * string) list)
      * Header.t)
      ref

    type read = bytes -> int -> int -> int

    let open_sync sync =
      (* Test wether the stream contains speex data *)
      let test_speex () =
        (* Get First page *)
        let page = Ogg.Sync.read sync in
        (* Check wether this is a b_o_s *)
        if not (Ogg.Page.bos page) then raise Not_found;
        (* Create a stream with this ID *)
        let serial = Ogg.Page.serialno page in
        let os = Ogg.Stream.create ~serial () in
        Ogg.Stream.put_page os page;
        (* Test header. Do not catch anything, first page should be sufficient *)
        let packet = Ogg.Stream.get_packet os in
        try
          let header = Header.header_of_packet packet in
          (* Get comments *)
          let page = ref (Ogg.Sync.read sync) in
          while Ogg.Page.serialno !page <> serial do
            page := Ogg.Sync.read sync
          done;
          Ogg.Stream.put_page os !page;
          let comments = Ogg.Stream.get_packet os in
          let comments = Header.comments_of_packet comments in
          (serial, os, header, comments)
        with _ ->
          Printf.printf "Not a speex stream..\n";
          raise Internal
      in
      let rec init () =
        try test_speex () with
          (* Not_found is not catched: ogg stream always start
             with all b_o_s and we don't care about sequenced streams here *)
          | Internal -> init ()
          | Ogg.Not_enough_data -> raise Not_speex
      in
      let serial, os, header, comments = init () in
      let chans = header.Header.nb_channels in
      let rate = header.Header.rate in
      let mode = header.Header.mode in
      let dec = Decoder.init mode in
      Decoder.set dec SPEEX_SET_SAMPLING_RATE rate;
      ref (dec, os, sync, serial, chans, comments, header)

    let open_file file =
      let sync, fd = Ogg.Sync.create_from_file file in
      (open_sync sync, fd)

    let open_feed feed =
      let sync = Ogg.Sync.create feed in
      open_sync sync

    let serial x =
      let _, _, _, serial, _, _, _ = !x in
      serial

    let comments x =
      let _, _, _, _, _, (_, l), _ = !x in
      l

    let header x =
      let _, _, _, _, _, _, h = !x in
      h

    let decode_gen f x =
      let dec, os, sync, serial, _, _, _ = !x in
      let dec = ref dec in
      let serial = ref serial in
      let os = ref os in
      let eos = ref false in
      let rec put () =
        try
          if !eos then (
            try
              (* Try to open a new stream *)
              let y = open_sync sync in
              let ndec, nos, _, nserial, _, _, _ = !y in
              x := !y;
              dec := ndec;
              serial := nserial;
              os := nos;
              eos := false
            with Not_found -> raise Internal);
          let page = Ogg.Sync.read sync in
          if Ogg.Stream.eos !os then eos := true;
          if Ogg.Page.serialno page = !serial then Ogg.Stream.put_page !os page
        with
          | Internal | Ogg.Not_enough_data -> raise Ogg.End_of_stream
          | Ogg.End_of_stream ->
              eos := true;
              put ()
      in
      let rec get () =
        try f !dec !os with
          | Ogg.End_of_stream ->
              eos := true;
              put ();
              get ()
          | Ogg.Not_enough_data ->
              put ();
              get ()
      in
      get ()

    let decode v = decode_gen Decoder.decode v
    let decode_stereo v = decode_gen Decoder.decode_stereo v

    let decode_feed v feed =
      decode_gen (fun x y -> Decoder.decode_feed x y feed) v

    let decode_feed_stereo v feed =
      decode_gen (fun x y -> Decoder.decode_feed_stereo x y feed) v

    let decode_int v = decode_gen Decoder.decode_int v
    let decode_int_stereo v = decode_gen Decoder.decode_int_stereo v

    let decode_int_feed v feed =
      decode_gen (fun x y -> Decoder.decode_int_feed x y feed) v

    let decode_int_feed_stereo v feed =
      decode_gen (fun x y -> Decoder.decode_int_feed_stereo x y feed) v
  end
end

module Skeleton = struct
  external fisbone :
    nativeint -> Header.t -> Int64.t -> string -> Ogg.Stream.packet
    = "ocaml_speex_skeleton_fisbone"

  let fisbone ?(start_granule = Int64.zero)
      ?(headers = [("Content-type", "audio/speex")]) ~serialno ~header () =
    let concat s (h, v) = Printf.sprintf "%s%s: %s\r\n" s h v in
    let s = List.fold_left concat "" headers in
    fisbone serialno header start_granule s
end
